File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/TransferAssetDefinitionInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/** Typed builder for the `TransferAssetDefinition` instruction. */
public final class TransferAssetDefinitionInstruction implements InstructionTemplate {

  private static final String ACTION = "TransferAssetDefinition";

  private final String sourceAccountId;
  private final String assetDefinitionId;
  private final String destinationAccountId;
  private final Map<String, String> arguments;

  private TransferAssetDefinitionInstruction(final Builder builder) {
    this(builder, builder.canonicalArguments());
  }

  private TransferAssetDefinitionInstruction(
      final Builder builder, final Map<String, String> argumentOrder) {
    this.sourceAccountId = builder.sourceAccountId;
    this.assetDefinitionId = builder.assetDefinitionId;
    this.destinationAccountId = builder.destinationAccountId;
    this.arguments =
        Collections.unmodifiableMap(new LinkedHashMap<>(Objects.requireNonNull(argumentOrder)));
  }

  public String sourceAccountId() {
    return sourceAccountId;
  }

  public String assetDefinitionId() {
    return assetDefinitionId;
  }

  public String destinationAccountId() {
    return destinationAccountId;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.TRANSFER;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static TransferAssetDefinitionInstruction fromArguments(final Map<String, String> arguments) {
    final Builder builder =
        builder()
            .setSourceAccountId(require(arguments, "source"))
            .setAssetDefinitionId(require(arguments, "definition"))
            .setDestinationAccountId(require(arguments, "destination"));
    return new TransferAssetDefinitionInstruction(builder, new LinkedHashMap<>(arguments));
  }

  private static String require(final Map<String, String> arguments, final String key) {
    final String value = arguments.get(key);
    if (value == null || value.isBlank()) {
      throw new IllegalArgumentException("Instruction argument '" + key + "' is required");
    }
    return value;
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof TransferAssetDefinitionInstruction other)) {
      return false;
    }
    return Objects.equals(sourceAccountId, other.sourceAccountId)
        && Objects.equals(assetDefinitionId, other.assetDefinitionId)
        && Objects.equals(destinationAccountId, other.destinationAccountId);
  }

  @Override
  public int hashCode() {
    return Objects.hash(sourceAccountId, assetDefinitionId, destinationAccountId);
  }

  public static final class Builder {
    private String sourceAccountId;
    private String assetDefinitionId;
    private String destinationAccountId;

    private Builder() {}

    public Builder setSourceAccountId(final String sourceAccountId) {
      this.sourceAccountId = Objects.requireNonNull(sourceAccountId, "sourceAccountId");
      return this;
    }

    public Builder setAssetDefinitionId(final String assetDefinitionId) {
      this.assetDefinitionId = Objects.requireNonNull(assetDefinitionId, "assetDefinitionId");
      return this;
    }

    public Builder setDestinationAccountId(final String destinationAccountId) {
      this.destinationAccountId =
          Objects.requireNonNull(destinationAccountId, "destinationAccountId");
      return this;
    }

    public TransferAssetDefinitionInstruction build() {
      if (sourceAccountId == null || sourceAccountId.isBlank()) {
        throw new IllegalStateException("sourceAccountId must be set");
      }
      if (assetDefinitionId == null || assetDefinitionId.isBlank()) {
        throw new IllegalStateException("assetDefinitionId must be set");
      }
      if (destinationAccountId == null || destinationAccountId.isBlank()) {
        throw new IllegalStateException("destinationAccountId must be set");
      }
      return new TransferAssetDefinitionInstruction(this);
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      args.put("source", sourceAccountId);
      args.put("definition", assetDefinitionId);
      args.put("destination", destinationAccountId);
      return args;
    }
  }
}

File: .taira-source-c80bf4d85a97/kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/TransferAssetDefinitionInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

/** Typed builder for the `TransferAssetDefinition` instruction. */
class TransferAssetDefinitionInstruction private constructor(
    val sourceAccountId: String,
    val assetDefinitionId: String,
    val destinationAccountId: String,
    override val arguments: Map<String, String>,
) : InstructionTemplate {

    override val kind: InstructionKind get() = InstructionKind.TRANSFER

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TransferAssetDefinitionInstruction) return false
        return sourceAccountId == other.sourceAccountId
            && assetDefinitionId == other.assetDefinitionId
            && destinationAccountId == other.destinationAccountId
    }

    override fun hashCode(): Int = listOf(sourceAccountId, assetDefinitionId, destinationAccountId).hashCode()

    class Builder internal constructor() {
        private var sourceAccountId: String? = null
        private var assetDefinitionId: String? = null
        private var destinationAccountId: String? = null

        fun setSourceAccountId(sourceAccountId: String) = apply {
            this.sourceAccountId = requireNotNull(sourceAccountId) { "sourceAccountId" }
        }

        fun setAssetDefinitionId(assetDefinitionId: String) = apply {
            this.assetDefinitionId = requireNotNull(assetDefinitionId) { "assetDefinitionId" }
        }

        fun setDestinationAccountId(destinationAccountId: String) = apply {
            this.destinationAccountId = requireNotNull(destinationAccountId) { "destinationAccountId" }
        }

        fun build(): TransferAssetDefinitionInstruction {
            val src = requireNotBlank(sourceAccountId, "sourceAccountId")
            val def = requireNotBlank(assetDefinitionId, "assetDefinitionId")
            val dst = requireNotBlank(destinationAccountId, "destinationAccountId")
            return TransferAssetDefinitionInstruction(src, def, dst, canonicalArguments(src, def, dst))
        }

        private fun canonicalArguments(src: String, def: String, dst: String): Map<String, String> =
            buildMap {
                put("action", ACTION)
                put("source", src)
                put("definition", def)
                put("destination", dst)
            }
    }

    companion object {
        private const val ACTION = "TransferAssetDefinition"

        @JvmStatic
        fun builder(): Builder = Builder()

        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): TransferAssetDefinitionInstruction {
            val src = require(arguments, "source")
            val def = require(arguments, "definition")
            val dst = require(arguments, "destination")
            return TransferAssetDefinitionInstruction(src, def, dst, LinkedHashMap(arguments))
        }

        private fun require(arguments: Map<String, String>, key: String): String {
            val value = arguments[key]
            require(!value.isNullOrBlank()) { "Instruction argument '$key' is required" }
            return value
        }

        private fun requireNotBlank(value: String?, name: String): String {
            check(!value.isNullOrBlank()) { "$name must be set" }
            return value
        }
    }
}
